internal/sessionmanager/sessionmanager.go

```diff
@@ -120,9 +122,9 @@ func (sm *SessionManager) ReceiveFrom(p peer.ID, blks []cid.Cid, haves []cid.Cid
 
 		if ok {
 			sess.ReceiveFrom(p, blks, haves, dontHaves)
-			sessions = append(sessions, sess)
 		}
 	}
 
-	return sessions
+	// Send CANCEL to all peers with want-have / want-block
+	sm.peerManager.SendCancels(ctx, blks)
```

Comment:
TODO: I think we might want to do this _first_, but I'm not sure. Let's not move it yet but remember to revisit this.

Reply:
Agreed, I think we can probably even do this in parallel to commiting the want to the blockstore